### PR TITLE
Yarn "For The Win"

### PR DESCRIPTION
**Pull request by bd82**
- Avoid duplicated Dev Deps (including CLI tools).
- Avoid multiple package-locks.json files
  - One yarn.lock to rule them all.
- No need for **2** step project init (`npm i && npm run bootstrap`)
  - Just execute `yarn` at root instead.

### CODE REVIEW

**Comment by matthew-dean** [comment]
@bd82 I'm curious about the removal of dev dependencies from individual packages. I guess the assumption is that tests will only be run from the git monorepo? That _may_ be a safe assumption, but....

_Usually_ a tenet of a monorepo is that if you were to move a package to an independent git repository, it would still 100% function as-is. And similarly, if a package were to be moved into the monorepo, it should run as-is without much refactoring.

It's also more self-documenting just which tests in those packages require which packages, vs. an implicit dependency of the higher-order mono-repo. So for now, I would prefer that packages have the appearance of independence, and all that `lerna`/`yarn` is doing is linking packages that happen to co-exist in the mono-repo (vs installing those other package dependencies via NPM).

**Comment by bd82** [comment]
> I'm curious about the removal of dev dependencies from individual packages. I guess the assumption is that tests will only be run from the git monorepo?

No, **you can run npm scripts from any sub package.**
`yarn run` wil resolve CLI tools from the root package as well as the current sub package
which is something `npm run` does not do...

> Usually a tenet of a monorepo is that if you were to move a package to an independent git repository, it would still 100% function as-is. And similarly, if a package were to be moved into the monorepo, it should run as-is without much refactoring.

> It's also more self-documenting just which tests in those packages require which packages, vs. an implicit dependency of the higher-order mono-repo. So for now, I would prefer that packages have the appearance of independence,

It is true some independence is lost, however 90% of the dev dependencies are **identical**...
If a package has a unique dev dep it should be listed at the sub package level. however if all packages
depend on some CLI (e.g mocha/typescript) than we avoid having **multiple versions** of those dev deps in the same mono repo.

Basically updating dev deps happens much more often than adding/removing packages
So its probably better to optimize for the common case rather than the theoretical edge case.

Another thing to consider i that the default behavior of lerna is a shared life cycle (releases/versions)
for all packages in a mono repo, so its not truly independent...

Basically it depends how you think about it.
- Do you want multiple fully independent packages that happen to be developed in the same repo.
- Or do you want a more modular structure for your **single** repo which happens to be constructed using npm packages?

**Comment by matthew-dean** [comment]
> Or do you want a more modular structure for your single project which happens to be constructed using npm packages?

It's probably more this one. If you think this more suitably matches the project, you have more knowledge in this domain, so I trust you. I just wanted to raise those concerns that it won't cause any maintainability headaches down the road, like harder for some future maintainer to reason about.

**Comment by matthew-dean** [comment]
@bd82 The one thing I'm considering is a scenario like a future extension of Less (like Jess) where I might want to pull packages and run certain test but...... maybe this wouldn't prevent that, since all the test data is installable.

**Comment by bd82** [comment]
> @bd82 The one thing I'm considering is a scenario like a future extension of Less (like Jess) where I might want to pull packages and run certain test but...... maybe this wouldn't prevent that, since all the test data is installable.

That is a unique scenario, maybe it could be modeled by creating "xxx-test" packages which
include their devDeps (e.g mocha) as real deps. so the test packages could be **npm installed**
and executed (given a programmatic API that accepts (dep injection) the implementation to be tested).

Maybe I'm describing a test conformance suite like test262?
- https://github.com/tc39/test262

Anyhow I think the scenario you are describing should be implemented with something
more advanced than git clone. and we can always duplicate/splitup the common later if needed...
